### PR TITLE
feat: Lean more on vue slots for state

### PR DESCRIPTION
**Pull request by richipargo**
Since Vue slots are useful to show different parts and states I'm leveraging on a better solution to how queries are loaded. Also covered some areas with unit testing

### CODE REVIEW

**Comment by richipargo** [comment]
Still have to document the changes

**Comment by paveltiunov** [comment]
@richipargo Hey Ricardo! This is big one! Thanks for contributing it! Could you please elaborate a little bit what problem does this PR solve?

**Comment by richipargo** [comment]
Hey @paveltiunov, more than a problem the intention of this PR is to give more flexibility on the usage of the `QueryBuilder` and `QueryRenderer` component. Using more `slots` to differentiate between the states of the query be it loaded, error or loading.
Also to try to cover the components a bit better I implemented Jest Unit testing on top of the components.

Still i'm missing a couple of items so I would leave it open still but comments and suggestions are welcome

**Comment by richipargo** [comment]
@paveltiunov It's ready!

**Comment by paveltiunov** [comment]
@richipargo Hey Ricardo! Cool! Are these changes backward compatible with previous version?

**Comment by richipargo** [comment]
Mostly yes

File: packages/cubejs-vue/src/QueryRenderer.js [hunk 1]
```diff
@@ -41,40 +44,47 @@ export default Vue.component('QueryRenderer', {
...
+      slot = $scopedSlots.default ? $scopedSlots.default(slotProps) : slot;
+    } else if (error) {
+      slot = $scopedSlots.error ? $scopedSlots.error({ error, sqlQuery }) : slot;
     }
+
+    return createElement(
+      'div',
+      {
+        class: {
+          'cubejs-query-renderer': true,
```
Comment (paveltiunov) [inline packages/cubejs-vue/src/QueryRenderer.js]: Do you use this one for styling?

**Comment by richipargo** [comment]
The idea is to make the object selectable if it requires styling, but of
course I do use it

**Comment by paveltiunov** [comment]
@richipargo Is there any chance we can make it fully backward compatible? I know some guys already deployed it. It would be great if we can make every version upgrade as smooth as possible.

**Comment by richipargo** [comment]
I wanted to avoid it cause it generated a lot of noise, but i've set back into the default slot the missing variables to mantain compatibility

File: packages/cubejs-vue/src/QueryRenderer.js [hunk 2]
```diff
@@ -41,40 +44,50 @@ export default Vue.component('QueryRenderer', {
...
-        this.$scopedSlots.default({
-          loading,
-          error,
-          sqlQuery,
-          ...this.builderProps,
-        }),
-      );
+      slot = $scopedSlots.default ? $scopedSlots.default(slotProps) : slot;
+    } else if (error) {
+      slot = $scopedSlots.error ? $scopedSlots.error({ error, sqlQuery }) : slot;
```
Comment (paveltiunov) [inline packages/cubejs-vue/src/QueryRenderer.js]: Do you think we should fallback here to default if only one slot is passed for backward compatibility?

Comment (richipargo) [reply packages/cubejs-vue/src/QueryRenderer.js]: You're right i missed that one

**Comment by paveltiunov** [comment]
@richipargo Hey Ricardo! Everything looks great! The only thing: is there any chance it can work without class names? Would be really great if we can maintain code to be class name free in order to prevent any class name clash.

**Comment by richipargo** [comment]
@paveltiunov while i would like to keep it I agree with the possible clashing, so I removed it

**Comment by paveltiunov** [comment]
@richipargo Cool! Thanks!

**Pull request merged**